docs/references/ic-interface-spec.md

```diff
@@ -1944,6 +1945,26 @@ The argument `type` decides which performance counter to return:
 
 In the future, the IC might expose more performance counters.
 
+### Canister limit {#system-api-canister-limit}
+
+The canister can query one of the "canister limits", which denotes a resource limit of the canister imposed by the current configuration of the IC and/or canister. This allows runtime logic, such garbage collector or persistence mechanism, to adapt to the system limits without recompilation and redeployment.
+
+`ic0.canister_limit : (limit_type : i32) -> i64`
+
+The argument `limit_type` determines which resource limit to return:
+
+-   0 : The [soft Wasm heap memory limit](https://internetcomputer.org/docs/current/developer-docs/smart-contracts/maintain/settings#wasm-memory-limit), or the IC-defined [hard Wasm memory limit](https://internetcomputer.org/docs/current/developer-docs/smart-contracts/maintain/resource-limits#resource-constraints-and-limits), whatever is smaller. The limit is specific to the canister, depending on the canister configuration (for the soft limit) and on whether it uses 32-bit or 64-bit Wasm memory.
+
+-   1 : The [instruction limit per update message](https://internetcomputer.org/docs/current/developer-docs/smart-contracts/maintain/settings#wasm-memory-limit).
+
+-   2 : The [instruction limit per upgrade message](https://internetcomputer.org/docs/current/developer-docs/smart-contracts/maintain/settings#wasm-memory-limit).
```

Comment:
We should explain in more detail what we mean here – sum of pre_upgrade and post_upgrade?